File: app/src/main/java/mega/privacy/android/app/FileDocument.java
package mega.privacy.android.app;

import java.io.File;

/*
 * File system document representation
 */
public class FileDocument {
    private File file;
    private MimeTypeList mimeType;
    private boolean isHighlighted;

    public FileDocument(File file, boolean isHighlighted) {
        this.file = file;
        this.isHighlighted = isHighlighted;
    }

    public File getFile() {
        return file;
    }

    public boolean isHighlighted() {
        return isHighlighted;
    }

    public boolean isHidden() {
        return getName().startsWith(".");
    }

    public boolean isFolder() {
        return file.isDirectory();
    }

    public long getSize() {
        return file.length();
    }

    public String getName() {
        return file.getName();
    }

    public MimeTypeList getMimeType() {
        if (mimeType == null) {
            mimeType = MimeTypeList.typeForName(getName());
        }
        return mimeType;
    }
}

File: app/src/main/java/mega/privacy/android/app/FileDocument.kt
package mega.privacy.android.app

import mega.privacy.android.data.extensions.toUri
import mega.privacy.android.domain.entity.document.DocumentEntity
import mega.privacy.android.domain.entity.uri.UriPath

/**
 * File system document representation
 * @param name
 * @param uriPath
 * @param isFolder
 * @param size
 * @param numFiles
 * @param numFolders
 * @param parent
 * @param canRead
 * @param isHighlighted
 *
 **/
data class FileDocument(
    val name: String,
    val uriPath: UriPath,
    val isFolder: Boolean,
    val size: Long,
    val numFiles: Int,
    val numFolders: Int = 0,
    val parent: FileDocument? = null,
    val canRead: Boolean = false,
    val isHighlighted: Boolean = false,
) {
    constructor(
        documentEntity: DocumentEntity,
        parent: FileDocument?,
        isHighlighted: Boolean = false,
    ) : this(
        name = documentEntity.name,
        uriPath = documentEntity.uri,
        isFolder = documentEntity.isFolder,
        size = documentEntity.size,
        numFiles = documentEntity.numFiles,
        numFolders = documentEntity.numFolders,
        parent = parent,
        canRead = documentEntity.canRead,
        isHighlighted = isHighlighted,
    )

    /**
     * total files and folders children
     */
    val totalChildren = numFiles + numFolders

    /**
     * If file is hidden, checking if name starts with `.`
     */
    val isHidden: Boolean get() = name.startsWith(".")

    /**
     * Gets the uri of this file (example `file:///path/filename.txt`)
     */
    fun getUri() = uriPath.toUri()
}
